src/instructlab/eval/gen_api_answer.py

```diff
@@ -0,0 +1,120 @@
+# Standard
+import concurrent.futures
+import json
+import os
+import time
+
+# Third Party
+# TODO need to look into this dependency
+from fastchat.model.model_adapter import get_conversation_template  # type: ignore
+import openai
+import shortuuid
+import tqdm
+
+# Local
+from .common import chat_completion_openai, load_questions, temperature_config
+
+
+def reorg_answer_file(answer_file):
+    """Sort by question id and de-duplication"""
+    answers = {}
+    with open(answer_file, "r", encoding="utf-8") as fin:
+        for l in fin:
+            qid = json.loads(l)["question_id"]
+            answers[qid] = l
+
+    qids = sorted(list(answers.keys()))
+    with open(answer_file, "w", encoding="utf-8") as fout:
+        for qid in qids:
+            fout.write(answers[qid])
+
+
+def get_answer(
+    question: dict,
+    model: str,
+    num_choices: int,
+    max_tokens: int,
+    answer_file: str,
+    force_temperature: float,
+):
+    assert force_temperature is None or question.get("required_temperature") is None
+    if force_temperature is not None:
+        temperature = force_temperature
+    elif "required_temperature" in question.keys():
+        temperature = question["required_temperature"]
+    elif question["category"] in temperature_config:
+        temperature = temperature_config[question["category"]]
+    else:
+        temperature = 0.7
+
+    choices = []
+    for i in range(num_choices):
+        conv = get_conversation_template(model)
+
+        turns = []
+        for j in range(len(question["turns"])):
+            conv.append_message(conv.roles[0], question["turns"][j])
+            conv.append_message(conv.roles[1], None)
+
+            output = chat_completion_openai(model, conv, temperature, max_tokens)
+
+            conv.update_last_message(output)
+            turns.append(output)
+
+        choices.append({"index": i, "turns": turns})
+
+    # Dump answers
+    ans = {
+        "question_id": question["question_id"],
+        "answer_id": shortuuid.uuid(),
+        "model_id": model,
+        "choices": choices,
+        "tstamp": time.time(),
+    }
+
+    os.makedirs(os.path.dirname(answer_file), exist_ok=True)
+    with open(answer_file, "a", encoding="utf-8") as fout:
+        fout.write(json.dumps(ans) + "\n")
+
+
+def run(
+    model_name,
+    question_begin=None,
+    question_end=None,
+    force_temperature=None,
+    answer_file=None,
+    num_choices=1,
+    max_tokens=1024,
+    parallel=1,
+    openai_api_base=None,
```

Comment:
Maybe we rename this arg to `judge_endpoint` just so that the user is not confused with the mention of `openai` in the arugment